pkg/controllers/namespace.go

```diff
@@ -0,0 +1,256 @@
+func (c *NamespaceConfig) handleDeleteNamespace(obj interface{}) {
+	ns, ok := obj.(*v1.Namespace)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+		}
+		if ns, ok = tombstone.Obj.(*v1.Namespace); !ok {
+			utilruntime.HandleError(fmt.Errorf("unexpected object type: %v", obj))
+			return
+		}
+	}
+	for i := range c.eventHandlers {
+		klog.V(4).Infof("Calling handler.OnNamespaceDelete")
+		c.eventHandlers[i].OnNamespaceDelete(ns)
+	}
+}
+
+// NamespaceInfo contains information that defines a namespace.
+type NamespaceInfo struct {
+	Name   string
+	Labels map[string]string
+}
+
+// nsChange represents a change in Namespace represented via NamespaceMap
+type nsChange struct {
+	previous NamespaceMap
+	current  NamespaceMap
+}
+
+// NamespaceChangeTracker carries state about uncommitted changes to an arbitrary number of
+// Namespaces in the node, keyed by their namespace and name
+type NamespaceChangeTracker struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items maps a service to its podChange.
+	items map[string]*nsChange
+}
+
+// newNamespaceInfo creates a new instance of NamespaceInfo
+func (nct *NamespaceChangeTracker) newNamespaceInfo(ns *v1.Namespace) *NamespaceInfo {
+	return &NamespaceInfo{
+		Name:   ns.Name,
+		Labels: ns.Labels,
+	}
+}
+
+// NewNamespaceChangeTracker creates a new instance of NamespaceChangeTracker
+func NewNamespaceChangeTracker() *NamespaceChangeTracker {
+	return &NamespaceChangeTracker{
+		items: make(map[string]*nsChange),
+	}
+}
+
+// nsToNamespaceMap creates NamespaceMap from v1.Namespace.
+// Note(adrianc): it is basically a map with single entry.
+func (nct *NamespaceChangeTracker) nsToNamespaceMap(ns *v1.Namespace) NamespaceMap {
+	if ns == nil {
+		return nil
+	}
+
+	namespaceMap := make(NamespaceMap)
+	nsInfo := nct.newNamespaceInfo(ns)
+	namespaceMap[ns.Name] = *nsInfo
+	return namespaceMap
+}
+
+// Update handles an update of a given Namespace
+func (nct *NamespaceChangeTracker) Update(previous, current *v1.Namespace) bool {
+	ns := current
+
+	if nct == nil {
+		return false
+	}
+
+	if ns == nil {
+		ns = previous
+	}
+	if ns == nil {
+		return false
+	}
+
+	nct.lock.Lock()
+	defer nct.lock.Unlock()
+
+	change, exists := nct.items[ns.Name]
+	if !exists {
+		change = &nsChange{}
+		prevNamespaceMap := nct.nsToNamespaceMap(previous)
+		change.previous = prevNamespaceMap
+		nct.items[ns.Name] = change
+	}
+	curNamespaceMap := nct.nsToNamespaceMap(current)
+	change.current = curNamespaceMap
+	if reflect.DeepEqual(change.previous, change.current) {
+		delete(nct.items, ns.Name)
+	}
+	return len(nct.items) >= 0
+}
+
+// NamespaceMap maps Namespace name to NamespaceInfo
+type NamespaceMap map[string]NamespaceInfo
+
+// Update updates podMap base on the given changes
+func (nm *NamespaceMap) Update(changes *NamespaceChangeTracker) {
+	if nm != nil {
```

Comment:
it can be if either previous or current is nil in Update call

when previous is nil and current is non-nil  -> object created
when previous is non-nil and current is nil -> object deleted
when previous and current are non nil -> object updated
when both are nil -> should not happen

Reply:
as discussed offline.

i will remove nil checks from changeTrackers as they are not needed IMO. 
for *Map objects lets keep them as the Update() logic of ChangeTracker relies on it.
i will fortify Map pointer receiver methods to check for nil across the board to be consistent although not all of them require it. I think it will help readability